perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_resource_attribute_def_virt_assignedGroups.java

```diff
@@ -0,0 +1,104 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Compute value of all groups assigned to the Resource and return list of their names.
+ * Name of group in the list is like '[VO_SHORT_NAME]:[PARENT_GROUP_NAME]:[GROUP_NAME]'
+ *
+ * If there is attribute 'resource:def:isAssignedWithSubgroups' set to true on the Resource, it will add all subgroups
+ * of assigned groups too.
+ *
+ * @author Michal Stava &lt;stavamichal@gmail.com&gt;
+ */
+@SkipValueCheckDuringDependencyCheck
+public class urn_perun_resource_attribute_def_virt_assignedGroups extends ResourceVirtualAttributesModuleAbstract implements ResourceVirtualAttributesModuleImplApi {
+
+	private final static Logger log = LoggerFactory.getLogger(urn_perun_resource_attribute_def_virt_assignedGroups.class);
+	private static final String A_R_isAssignedWithSubgroups = AttributesManager.NS_RESOURCE_ATTR_DEF + ":isAssignedWithSubgroups";
+
+	@Override
+	public Attribute getAttributeValue(PerunSessionImpl sess, Resource resource, AttributeDefinition attributeDefinition) throws InternalErrorException {
+		Attribute attribute = new Attribute(attributeDefinition);
+		Set<String> allUniqueGroupNames = new HashSet<>();
+
+		Vo vo = sess.getPerunBl().getResourcesManagerBl().getVo(sess, resource);
+		List<Group> assignedGroups = sess.getPerunBl().getResourcesManagerBl().getAssignedGroups(sess, resource);
+
+		boolean isAssignedWithSubgroups;
+		try {
+			Attribute isAssignedWithSubgroupsAttribute = sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, resource, A_R_isAssignedWithSubgroups);
+			if(isAssignedWithSubgroupsAttribute.getValue() == null) {
+				isAssignedWithSubgroups = false;
+			} else {
+				isAssignedWithSubgroups = isAssignedWithSubgroupsAttribute.valueAsBoolean();
+			}
+		} catch (AttributeNotExistsException ex) {
+			log.debug("There is missing definition of attribute {}", A_R_isAssignedWithSubgroups);
+			isAssignedWithSubgroups = false;
+		} catch (WrongAttributeAssignmentException ex) {
+			throw new InternalErrorException(ex);
+		}
+
+		for(Group assignedGroup: assignedGroups) {
+			allUniqueGroupNames.add(computeGroupName(vo, assignedGroup));
+			if(isAssignedWithSubgroups) {
+				for(Group subgroup: sess.getPerunBl().getGroupsManagerBl().getAllSubGroups(sess, assignedGroup)) {
+					allUniqueGroupNames.add(computeGroupName(vo, subgroup));
+				}
+			}
+		}
+
+		attribute.setValue(new ArrayList(allUniqueGroupNames));
+
+		return attribute;
+	}
+
+	/**
+	 * Return group name like '[VO_SHORT_NAME]:[GROUP_NAME]'
+	 *
+	 * @param vo to get short name from
+	 * @param group to get name from
+	 * @return computed group name
+	 */
+	private String computeGroupName(Vo vo, Group group) {
+		return vo.getShortName() + ":" + group.getName();
+	}
+
+	@Override
+	public List<String> getStrongDependencies() {
+		List<String> strongDependencies = new ArrayList<>();
+		strongDependencies.add(AttributesManager.NS_VO_ATTR_CORE + ":shortName");
```

Comment:
Why ? We retrieve value from the object ``Vo`` and do not call ``getAttribute()`` on such core attribute.

Reply:
Copy&Paste mistake. Fixed. :)